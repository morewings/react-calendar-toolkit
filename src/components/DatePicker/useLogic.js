import {useCallback} from 'react';
import {
  useDatePickerActions,
  useDatePickerContext,
  useIncrementPrecision,
} from 'features/datepicker';

const useLogic = (onDateSet, minPrecision) => {
  const {
    state: {precision},
  } = useDatePickerContext();
  const {setDate, setVisibility} = useDatePickerActions();
  const incrementPrecision = useIncrementPrecision(minPrecision);
  const handleDateSet = useCallback(
    date => {
      if (precision === minPrecision) {
        onDateSet(date);
        setDate(date);
      } else {
        setVisibility(date);
        incrementPrecision();
      }
    },
    [
      incrementPrecision,
      minPrecision,
      onDateSet,
      precision,
      setDate,
      setVisibility,
    ]
  );
  return {handleDateSet};
};

export default useLogic;
