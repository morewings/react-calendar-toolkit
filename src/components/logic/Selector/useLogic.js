import {useCallback} from 'react';
import {
  ceilMonth,
  checkIsWithinInterval,
  decrementMonth,
  floorMonth,
  incrementMonth,
} from 'utils/dateUtils';
import {useDatePickerActions} from 'features/datepicker';

const useLogic = ({startDate, endDate, visibleTimestamp}) => {
  const {setVisibility, setPrecision} = useDatePickerActions();

  const isAvailable = useCallback(
    date =>
      checkIsWithinInterval(
        {start: floorMonth(startDate), end: ceilMonth(endDate)},
        date
      ),
    [endDate, startDate]
  );

  const onSetPrecision = useCallback(
    nextPrecision => {
      setPrecision(nextPrecision);
    },
    [setPrecision]
  );

  const onIncrementMonth = useCallback(() => {
    const nextDate = incrementMonth(visibleTimestamp, 1);
    isAvailable(nextDate) && setVisibility(nextDate);
  }, [isAvailable, setVisibility, visibleTimestamp]);

  const onDecrementMonth = useCallback(() => {
    const nextDate = decrementMonth(visibleTimestamp, 1);
    isAvailable(nextDate) && setVisibility(nextDate);
  }, [isAvailable, setVisibility, visibleTimestamp]);

  return {onSetPrecision, onIncrementMonth, onDecrementMonth};
};

export default useLogic;
