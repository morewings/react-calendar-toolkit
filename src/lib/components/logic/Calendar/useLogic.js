import {useCallback} from 'react';
import {
  ceilMonth,
  checkIsSameMonth,
  checkIsWeekend,
  checkIsWithinInterval,
  floorMonth,
  matchDatesWithPrecision,
} from 'lib/utils/dateUtils';

const useLogic = ({
  disableDate,
  endDate,
  precision,
  startDate,
  highlightDate,
  onDateSet,
  highlightWeekends,
  visibleTimestamp,
  selectedTimestamp,
  todayTimestamp,
}) => {
  const getIsDisabled = useCallback(
    date => {
      if (precision === 'year') return false; // we are not disabling years, just rendering the range
      if (precision === 'month') {
        return !checkIsWithinInterval(
          {start: floorMonth(startDate), end: ceilMonth(endDate)},
          date
        );
      }
      return (
        disableDate({date, isWeekend: checkIsWeekend(date), precision}) ||
        !checkIsWithinInterval({start: startDate, end: endDate}, date)
      );
    },
    [disableDate, endDate, precision, startDate]
  );

  const getIsHighlighted = useCallback(
    date => highlightDate({date, isWeekend: checkIsWeekend(date), precision}),
    [highlightDate, precision]
  );

  const handleDateSet = useCallback(
    date => {
      const isDisabled = getIsDisabled(date);
      !isDisabled && onDateSet(date);
    },
    [getIsDisabled, onDateSet]
  );

  const getIsWeekend = date =>
    precision === 'day' && highlightWeekends && checkIsWeekend(date);

  const getBelongsToSameMonth = date =>
    precision === 'day' && checkIsSameMonth(date, visibleTimestamp);

  const getIsSelected = date =>
    matchDatesWithPrecision(precision, date, selectedTimestamp);

  const getIsCurrent = date =>
    matchDatesWithPrecision(precision, date, todayTimestamp);

  return {
    getIsDisabled,
    getIsHighlighted,
    handleDateSet,
    getIsWeekend,
    getBelongsToSameMonth,
    getIsSelected,
    getIsCurrent,
  };
};

export default useLogic;
