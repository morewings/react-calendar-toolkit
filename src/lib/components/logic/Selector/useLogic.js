import {useCallback} from 'react';
import {
  ceilMonth,
  checkIsWithinInterval,
  subMonth,
  floorMonth,
  addMonth,
} from 'lib/utils/dateUtils';
import {useDatepickerActions} from 'lib/features/datepicker';

const useLogic = ({startDate, endDate, visibleTimestamp}) => {
  const {setVisibility, setPrecision} = useDatepickerActions();

  const isAvailable = useCallback(
    date =>
      checkIsWithinInterval(
        {start: floorMonth(startDate), end: ceilMonth(endDate)},
        date
      ),
    [endDate, startDate]
  );

  const onSetPrecision = useCallback(
    nextPrecision => {
      setPrecision(nextPrecision);
    },
    [setPrecision]
  );

  const onIncrementMonth = useCallback(() => {
    const nextDate = addMonth(visibleTimestamp, 1);
    isAvailable(nextDate) && setVisibility(nextDate);
  }, [isAvailable, setVisibility, visibleTimestamp]);

  const onDecrementMonth = useCallback(() => {
    const nextDate = subMonth(visibleTimestamp, 1);
    isAvailable(nextDate) && setVisibility(nextDate);
  }, [isAvailable, setVisibility, visibleTimestamp]);

  return {onSetPrecision, onIncrementMonth, onDecrementMonth};
};

export default useLogic;
