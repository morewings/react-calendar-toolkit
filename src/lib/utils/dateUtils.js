import {
  formatWithOptions,
  isSameMonth,
  isSameYear,
  toDate,
  getTime,
  addMonths,
  addDays,
  addYears,
  isWithinInterval,
  isWeekend,
  isSameDay,
  isSameHour,
  getDate,
  startOfMonth,
  endOfMonth,
} from 'date-fns/fp';
import compose from 'lib/utils/compose';

/**
 * Unix timestamp
 * @typedef {number} UnixTimestamp
 */

/**
 * Date union type may be Date or Unix timestamp
 * @typedef {(Date|UnixTimestamp)} DateUnion
 */

/**
 * @function
 * @name formatDateWithLocale
 * @description Returns formatted date
 * @param {Object} options - Options object with locale
 * @param {string} pattern - Formatting pattern
 * @param {Date} date - Date object to apply format
 * @return {string} Formatted date
 */
export const formatDateWithLocale = formatWithOptions;

/**
 * @function
 * @name checkIsSameMonth
 * @description Checks if two dates belong to the same month
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameMonth = isSameMonth;

/**
 * @function
 * @name checkIsSameYear
 * @description Checks if two dates belong to the same year
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameYear = isSameYear;

/**
 * @function
 * @name checkIsSameDay
 * @description Checks if two dates belong to the same day
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameDay = isSameDay;

/**
 * @function
 * @name checkIsSameHour
 * @description Checks if two dates belong to the same hour
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const checkIsSameHour = isSameHour;

/**
 * @function
 * @name convertToTimestamp
 * @description Converts provided date to Unix timestamp
 * @param {Date} date - Date to convert
 * @return {UnixTimestamp} Unix timestamp
 */
export const convertToTimestamp = date => getTime(date);

/**
 * @function
 * @name convertToDate
 * @description Converts provided Unix timestamp to Date object
 * @param {number} timestamp - Date to convert
 * @return {Date}
 */
export const convertToDate = timestamp => toDate(timestamp);

/**
 * @function
 * @name convertToDayOfMonth
 * @description Converts provided Unix timestamp or Date to day of month number
 * @param {DateUnion} timestamp - Timestamp or Date to convert
 * @return {number} Day of the month
 */
export const convertToDayOfMonth = timestamp => getDate(timestamp);

/**
 * @function
 * @name checkIsWithinInterval
 * @description Checks if date belongs to the interval
 * @typedef {Object} Interval
 * @property {Date} start - Start of interval
 * @property {Date} end - End of interval
 * @param {DateUnion} date - Date to check
 * @param {Interval} interval - Time range to check
 * @return {boolean}
 */
export const checkIsWithinInterval = isWithinInterval;

/**
 * @function
 * @name checkIsWeekend
 * @description Checks if date belongs to the weekend
 * @param {DateUnion} date - Date to check
 * @return {boolean}
 */
export const checkIsWeekend = date => isWeekend(date);

/**
 * @function
 * @name matchDatesWithPrecision
 * @description Checks if dates are the same with given precision, eg belong to same day|month|year
 * @param {string} precision - Date to check
 * @param {DateUnion} dateLeft - Date to check
 * @param {DateUnion} dateRight - Date to check
 * @return {boolean}
 */
export const matchDatesWithPrecision = (precision, dateLeft, dateRight) =>
  ({
    day: checkIsSameDay,
    month: checkIsSameMonth,
    year: checkIsSameYear,
  }[precision](dateLeft, dateRight));

/**
 * @function
 * @name ceilMonth
 * @description Gets end of month date
 * @param {DateUnion} date
 * @return {Date}
 */
export const ceilMonth = endOfMonth;

/**
 * @function
 * @name floorMonth
 * @description Gets start of month date
 * @param {DateUnion} date
 * @return {Date}
 */
export const floorMonth = startOfMonth;

/**
 * @function
 * @name addMonth
 * @description Adds months
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of months to add
 * @return {UnixTimestamp}
 */
export const addMonth = (date, amount) =>
  compose(convertToTimestamp, addMonths)(amount, date);

/**
 * @function
 * @name subMonth
 * @description Subtracts months
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of months to subtract
 * @return {UnixTimestamp}
 */
export const subMonth = (date, amount) =>
  compose(convertToTimestamp, addMonths)(-amount, date);

/**
 * @function
 * @name addDay
 * @description Adds days
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to add
 * @return {UnixTimestamp}
 */
export const addDay = (date, amount) =>
  compose(convertToTimestamp, addDays)(amount, date);

/**
 * @function
 * @name subDay
 * @description Subtracts days
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to subtract
 * @return {UnixTimestamp}
 */
export const subDay = (date, amount) =>
  compose(convertToTimestamp, addDays)(-amount, date);

/**
 * @function
 * @name addYear
 * @description Adds years
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of years to add
 * @return {UnixTimestamp}
 */
export const addYear = (date, amount) =>
  compose(convertToTimestamp, addYears)(amount, date);

/**
 * @function
 * @name subYear
 * @description Subtracts years
 * @param {DateUnion} date - Date or Unix timestamp
 * @param {number} amount - Amount of days to subtract
 * @return {UnixTimestamp}
 */
export const subYear = (date, amount) =>
  compose(convertToTimestamp, addYears)(-amount, date);
