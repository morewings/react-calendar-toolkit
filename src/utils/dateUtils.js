import locale from 'date-fns/esm/locale/ru';
import {
  addDays,
  differenceInDays,
  endOfMonth,
  endOfWeek,
  startOfMonth,
  startOfWeek,
  getDate,
  isSameMonth,
  isSameDay,
  format,
  addMonths,
  getTime,
  toDate,
  startOfYear,
} from 'date-fns';

export const getWeekDayNames = () =>
  [...Array(7).keys()].map(i => locale.localize.day(i, {width: 'short'}));

export const getMonthNames = () =>
  [...Array(12).keys()].map(i => locale.localize.month(i, {width: 'wide'}));

export const getMonths = date => {
  const year = startOfYear(date);
  return getMonthNames().map((name, i) => ({
    name,
    date: toDate(getAddMonth(year, i)), // TODO: fix this shit, double date conversion
  }));
};

export const getMonthDays = timestamp => {
  const monthStart = startOfMonth(timestamp);
  const monthEnd = endOfMonth(monthStart);
  const startDate = startOfWeek(monthStart);
  const endDate = endOfWeek(monthEnd);
  const cellLength = differenceInDays(endDate, startDate);
  return new Array(cellLength + 1).fill('').map((_, i) => ({
    date: addDays(startDate, i),
    dayNumber: getDate(addDays(startDate, i)),
  }));
};

export const getIsSameMonth = (day, month) => {
  const monthStart = startOfMonth(month);
  return isSameMonth(day, monthStart);
};

export const getIsSameDay = (dateLeft, dateRight) =>
  isSameDay(dateLeft, dateRight);

export const getFormattedDate = (template, date) => format(date, template);

export {getUnixTime} from 'date-fns';

export {getTime, toDate};

export const getAddMonth = (date, amount) =>
  getTime(addMonths(toDate(date), amount)); // TODO: fix this double date

export const getSubMonth = (date, amount) => getAddMonth(date, -amount);
