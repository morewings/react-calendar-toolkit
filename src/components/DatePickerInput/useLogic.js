import {useCallback} from 'react';
import {useModalActions} from 'features/modal';
import {convertToDate} from 'utils/dateUtils';
import {useFormatDate} from 'features/locale';
import {useDatePickerContext} from 'features/datepicker';

const useLogic = ({onDateSet, hideOnSelect}) => {
  const {toggleDatePicker} = useModalActions();
  const formatDate = useFormatDate();
  const {
    state: {selectedTimestamp},
  } = useDatePickerContext();

  const handleDateSet = useCallback(
    date => {
      onDateSet(date);
      hideOnSelect && toggleDatePicker(false);
    },
    [hideOnSelect, onDateSet, toggleDatePicker]
  );

  const formatValue = useCallback(
    formatPattern =>
      formatDate(formatPattern, convertToDate(selectedTimestamp)),
    [formatDate, selectedTimestamp]
  );

  const getDate = () => convertToDate(selectedTimestamp);

  return {handleDateSet, formatValue, getDate};
};

export default useLogic;
