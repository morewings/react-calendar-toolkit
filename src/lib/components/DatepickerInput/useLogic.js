import {useCallback} from 'react';
import {useModalActions} from 'lib/features/modal';
import {convertToDate} from 'lib/utils/dateUtils';
import {useFormatDate} from 'lib/features/locale';
import {useDatepickerContext} from 'lib/features/datepicker';

const useLogic = ({onDateSet, hideOnSelect}) => {
  const {toggleDatepicker} = useModalActions();
  const formatDate = useFormatDate();
  const {
    state: {selectedTimestamp},
  } = useDatepickerContext();

  const handleDateSet = useCallback(
    date => {
      onDateSet(date);
      hideOnSelect && toggleDatepicker(false);
    },
    [hideOnSelect, onDateSet, toggleDatepicker]
  );

  const formatValue = useCallback(
    formatPattern =>
      formatDate(formatPattern, convertToDate(selectedTimestamp)),
    [formatDate, selectedTimestamp]
  );

  const getDate = () => convertToDate(selectedTimestamp);

  return {handleDateSet, formatValue, getDate};
};

export default useLogic;
