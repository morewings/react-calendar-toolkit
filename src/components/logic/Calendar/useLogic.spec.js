import {renderHook, act} from '@testing-library/react-hooks';
import {checkIsSameDay} from 'utils/dateUtils';
import useLogic from './useLogic';

const today = new Date(2011, 10, 11).getTime(); // 11.11.2011
const yesterday = new Date(2011, 10, 10).getTime(); // 10.11.2011
const matchYesterday = jest.fn(
  ({date, precision}) => precision === 'day' && checkIsSameDay(date, yesterday)
);
const tomorrow = new Date(2011, 10, 12).getTime(); // 12.11.2011
const nextMonth = new Date(2011, 11, 11).getTime(); // 11.12.2011
const prevMonth = new Date(2011, 9, 11).getTime(); // 11.10.2011
const nextYear = new Date(2012, 10, 11).getTime(); // 11.11.2012
const prevYear = new Date(2010, 10, 11).getTime(); // 11.11.2010
const startYear = new Date(2011, 0, 1); // 01.01.2011
const startMonth = new Date(2011, 10, 1); // 01.11.2011
const endYear = new Date(2011, 11, 31); // 31.12.2011
const endMonth = new Date(2011, 10, 30); // 30.11.2011

const weekend = tomorrow;
const notWeekend = today;

const selectedDay = yesterday;
const notSelectedDay = tomorrow;
const selectedMonth = startMonth;
const notSelectedMonth = nextMonth;
const selectedYear = startYear;
const notSelectedYear = prevYear;

const matchTrue = jest.fn(() => true);

const defaultProps = {
  onDateSet: jest.fn(),
  todayTimestamp: today,
  selectedTimestamp: yesterday,
  visibleTimestamp: tomorrow,
  precision: 'day',
  startDate: startYear,
  endDate: endYear,
  disableDate: jest.fn(),
  highlightDate: jest.fn(),
  highlightWeekends: true,
};

describe('Calendar > useLogic ', () => {
  beforeEach(() => {
    defaultProps.onDateSet.mockClear();
    matchYesterday.mockClear();
    matchTrue.mockClear();
  });

  it('returns functions', () => {
    const {result} = renderHook(() => useLogic(defaultProps));
    expect(result.current).toMatchSnapshot();
  });

  describe('getIsDisabled', () => {
    it('returns false, if precision === `year`', () => {
      const props = {
        ...defaultProps,
        precision: 'year',
      };
      const {result} = renderHook(() => useLogic(props));
      expect(result.current.getIsDisabled(today)).toBe(false);
    });

    describe.each`
      precision  | date         | start         | end         | matcher           | expected
      ${'month'} | ${today}     | ${startYear}  | ${endYear}  | ${undefined}      | ${false}
      ${'month'} | ${nextYear}  | ${startYear}  | ${endYear}  | ${undefined}      | ${true}
      ${'month'} | ${prevYear}  | ${startYear}  | ${endYear}  | ${undefined}      | ${true}
      ${'day'}   | ${today}     | ${startMonth} | ${endMonth} | ${undefined}      | ${false}
      ${'day'}   | ${yesterday} | ${startMonth} | ${endMonth} | ${matchYesterday} | ${true}
      ${'day'}   | ${nextMonth} | ${startMonth} | ${endMonth} | ${matchYesterday} | ${true}
      ${'day'}   | ${prevMonth} | ${startMonth} | ${endMonth} | ${matchYesterday} | ${true}
    `(
      `checks if date is within provided range,`,
      ({precision, date, start, end, expected, matcher}) => {
        it(` if precision === ${precision}`, () => {
          const initialProps = {
            ...defaultProps,
            precision,
            startDate: start,
            endDate: end,
            expected,
            disableDate: matcher || defaultProps.disableDate,
          };
          const {result} = renderHook(props => useLogic(props), {
            initialProps,
          });
          expect(result.current.getIsDisabled(date)).toBe(expected);
          if (matcher) {
            expect(matcher).toHaveBeenCalledTimes(1);
            expect(matcher.mock.calls[0]).toMatchSnapshot();
          }
        });
      }
    );
  });

  describe.each`
    precision  | date        | matcher
    ${'year'}  | ${today}    | ${matchTrue}
    ${'month'} | ${today}    | ${matchTrue}
    ${'day'}   | ${today}    | ${matchTrue}
    ${'day'}   | ${tomorrow} | ${matchTrue}
  `('getIsHighlighted', ({precision, date, matcher}) => {
    it(`invokes matcher with arguments`, () => {
      const initialProps = {
        ...defaultProps,
        precision,
        highlightDate: matcher,
      };
      const {result} = renderHook(props => useLogic(props), {
        initialProps,
      });
      expect(result.current.getIsHighlighted(date)).toBe(true);
      expect(matcher).toHaveBeenCalledTimes(1);
      expect(matcher.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('handleDateSet', () => {
    it('calls `onDateSet` callback with date, if it is not disabled', () => {
      const {result} = renderHook(props => useLogic(props), {
        initialProps: defaultProps,
      });
      act(() => {
        result.current.handleDateSet(today);
      });
      expect(defaultProps.onDateSet).toHaveBeenCalledTimes(1);
      expect(defaultProps.onDateSet).toHaveBeenCalledWith(today);
    });

    it('does not call `onDateSet` callback with date, if it is disabled', () => {
      const {result} = renderHook(props => useLogic(props), {
        initialProps: {
          ...defaultProps,
          disableDate: ({date}) => date === today,
        },
      });
      act(() => {
        result.current.handleDateSet(today);
      });
      expect(defaultProps.onDateSet).not.toHaveBeenCalled();
    });

    describe.each`
      precision  | date          | highlightWeekends | expected
      ${'day'}   | ${notWeekend} | ${true}           | ${false}
      ${'day'}   | ${weekend}    | ${true}           | ${true}
      ${'day'}   | ${weekend}    | ${false}          | ${false}
      ${'month'} | ${weekend}    | ${true}           | ${false}
      ${'year'}  | ${weekend}    | ${true}           | ${false}
    `('getIsWeekend', ({precision, date, highlightWeekends, expected}) => {
      it('checks if date is weekend', () => {
        const {result} = renderHook(props => useLogic(props), {
          initialProps: {
            ...defaultProps,
            precision,
            highlightWeekends,
          },
        });
        expect(result.current.getIsWeekend(date)).toBe(expected);
      });
    });
  });

  describe.each`
    precision  | date         | expected
    ${'day'}   | ${yesterday} | ${true}
    ${'day'}   | ${nextMonth} | ${false}
    ${'month'} | ${yesterday} | ${false}
    ${'year'}  | ${yesterday} | ${false}
  `('getBelongsToSameMonth', ({precision, date, expected}) => {
    it('checks if day entry belongs to this calendar view', () => {
      const {result} = renderHook(props => useLogic(props), {
        initialProps: {
          ...defaultProps,
          precision,
        },
      });
      expect(result.current.getBelongsToSameMonth(date)).toBe(expected);
    });
  });

  describe.each`
    precision  | date                | expected
    ${'day'}   | ${selectedDay}      | ${true}
    ${'day'}   | ${notSelectedDay}   | ${false}
    ${'month'} | ${selectedDay}      | ${true}
    ${'month'} | ${notSelectedDay}   | ${true}
    ${'month'} | ${selectedMonth}    | ${true}
    ${'month'} | ${notSelectedMonth} | ${false}
    ${'year'}  | ${selectedDay}      | ${true}
    ${'year'}  | ${notSelectedDay}   | ${true}
    ${'year'}  | ${selectedMonth}    | ${true}
    ${'year'}  | ${notSelectedMonth} | ${true}
    ${'year'}  | ${selectedYear}     | ${true}
    ${'year'}  | ${notSelectedYear}  | ${false}
  `('getIsSelected', ({precision, date, expected}) => {
    it('checks if this date is selected using precision', () => {
      const {result} = renderHook(props => useLogic(props), {
        initialProps: {
          ...defaultProps,
          precision,
        },
      });
      expect(result.current.getIsSelected(date)).toBe(expected);
    });
  });

  describe.each`
    precision  | date                | expected
    ${'day'}   | ${today}            | ${true}
    ${'day'}   | ${yesterday}        | ${false}
    ${'month'} | ${today}            | ${true}
    ${'month'} | ${yesterday}        | ${true}
    ${'month'} | ${selectedMonth}    | ${true}
    ${'month'} | ${notSelectedMonth} | ${false}
    ${'year'}  | ${today}            | ${true}
    ${'year'}  | ${yesterday}        | ${true}
    ${'year'}  | ${selectedMonth}    | ${true}
    ${'year'}  | ${notSelectedMonth} | ${true}
    ${'year'}  | ${selectedYear}     | ${true}
    ${'year'}  | ${notSelectedYear}  | ${false}
  `('getIsCurrent', ({precision, date, expected}) => {
    it('checks if this date is current using precision', () => {
      const {result} = renderHook(props => useLogic(props), {
        initialProps: {
          ...defaultProps,
          precision,
        },
      });
      expect(result.current.getIsCurrent(date)).toBe(expected);
    });
  });
});
