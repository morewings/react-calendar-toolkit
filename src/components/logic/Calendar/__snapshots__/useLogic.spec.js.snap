// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar > useLogic  getIsDisabled checks if date is within provided range,  if precision === day 1`] = `
Array [
  Object {
    "date": 1320883200000,
    "isWeekend": false,
    "precision": "day",
  },
]
`;

exports[`Calendar > useLogic  getIsDisabled checks if date is within provided range,  if precision === day 2`] = `
Array [
  Object {
    "date": 1323561600000,
    "isWeekend": true,
    "precision": "day",
  },
]
`;

exports[`Calendar > useLogic  getIsDisabled checks if date is within provided range,  if precision === day 3`] = `
Array [
  Object {
    "date": 1318291200000,
    "isWeekend": false,
    "precision": "day",
  },
]
`;

exports[`Calendar > useLogic  getIsHighlighted invokes matcher with arguments 1`] = `
Array [
  Object {
    "date": 1320969600000,
    "isWeekend": false,
    "precision": "year",
  },
]
`;

exports[`Calendar > useLogic  getIsHighlighted invokes matcher with arguments 2`] = `
Array [
  Object {
    "date": 1320969600000,
    "isWeekend": false,
    "precision": "month",
  },
]
`;

exports[`Calendar > useLogic  getIsHighlighted invokes matcher with arguments 3`] = `
Array [
  Object {
    "date": 1320969600000,
    "isWeekend": false,
    "precision": "day",
  },
]
`;

exports[`Calendar > useLogic  getIsHighlighted invokes matcher with arguments 4`] = `
Array [
  Object {
    "date": 1321056000000,
    "isWeekend": true,
    "precision": "day",
  },
]
`;

exports[`Calendar > useLogic  returns functions 1`] = `
Object {
  "getBelongsToSameMonth": [Function],
  "getIsCurrent": [Function],
  "getIsDisabled": [Function],
  "getIsHighlighted": [Function],
  "getIsSelected": [Function],
  "getIsWeekend": [Function],
  "handleDateSet": [Function],
}
`;
